site/sfguides/src/getting_started_with_traces/getting_started_with_traces.md

```diff
@@ -0,0 +1,445 @@
+
+<!-- ------------------------ -->
+## Enabling Traces
+Duration: 2
+
+In order to get logs, errors and traces, levels need to be modified. This can easily be done in sql.
+
+Change the third notebook cell to SQL and paste this section into the body.
+
+```sql
+ALTER DATABASE "{{database_name}}" SET LOG_LEVEL = DEBUG;
+ALTER SCHEMA "{{schema_name}}" SET LOG_LEVEL = DEBUG;
+
+ALTER DATABASE "{{database_name}}" SET TRACE_LEVEL = ALWAYS;
+ALTER SCHEMA "{{schema_name}}" SET TRACE_LEVEL = ALWAYS;
+
+ALTER DATABASE "{{database_name}}" SET METRIC_LEVEL = ALL;
+ALTER SCHEMA "{{schema_name}}" SET METRIC_LEVEL = ALL;
+
+ALTER SESSION SET METRIC_LEVEL = ALL;
+
+```
+
+Run this cell and verify it is successful.
+
+<!-- ------------------------ -->
+## Create the UDF
+Duration: 5
+
+A UDF will use the External Access and Secret created previously and pull data from the OpenWeather API using Python Requests library. It will pull the current weather for a specific latitude and longitude.
+
+In order to populate the bronze layer, a stored procedure will call the UDF for every city in the city_list to get the current weather. This is done with a dataframe and will append the data to a table for usage in the future.
+
+When calling to a 3rd party API like this, it is important to know the performance and potential errors coming from those calls. To trace this information, the function uses a custom span in OpenTelemetry. Review the tracer variable and how that is used. Without the custom span, it would be difficult to know which calls were erroring out to the API and what status code the API was returning.
+
+The function will be stored in the database schema and can be used in the future from sql or python, as we will do in the next cell.
+
+Create a new python cell at the bottom of the notebook by clicking on + Python.
+
+Paste this section into the new cell.
+
+```python
+import _snowflake
+import requests
+import json
+import logging
+
+from opentelemetry import trace
+from snowflake import telemetry
+
+rsession = requests.Session()
+def get_weather(lat, lon):
+  """
+  A UDF handler. Given a latitude and longitude, this UDF will use external network access to get the 
+  weather for that area from the OpenWeatherMap API.
+  """
+  api_key = _snowflake.get_generic_secret_string('api_key')
+  url = f"https://api.openweathermap.org/data/2.5/weather?lat={lat}&lon={lon}&exclude=hourly,daily&appid={api_key}"
+  tracer = trace.get_tracer(__name__) 
+  with tracer.start_as_current_span(f"openweather_api_call") as p:
+    response = rsession.get(url)
+    logging.debug(f"Body from API: {response.text} in get_weather")
+    if response.status_code == 200:
+        p.set_status(trace.Status(trace.StatusCode.OK))
+    else:
+        p.set_status(trace.Status(trace.StatusCode.ERROR, str(response.status_code)))
+        logging.warn(f"Unexpected response from API: {response.status_code} in get_weather")
+  return response.json()
```

Comment:
I think we should refactor this to use span events or debug-level logging in the UDF since the UI isn't representing it correctly. Let me schedule a quick call with us and the telemetry team about this scenario, but in any case I don't think it will be fixed before the 15th so we should think about alternatives.

We can add custom spans in the "while" loop of the stored proc to showcase custom spans instead

Reply:
I agree this scenario was not tested exhaustively, and there could be bugs related to this case. +1 on using span events in the Quick Start, so that this tutorial is not blocked by potential bug fixes needed.